### PR TITLE
next: Add Use and support for parsing use statements

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
next: Add Use and support for parsing use statements

This is for `compiler/next`.

Add support for parsing use statements. Represent use statements via
a combination of three nodes:

  - As
  - Use
  - UseClause

A use statement is represented as a `Use` node in the `uAST`. It
contains a visibility (e.g. public/private) along with one or more
use clauses of type `UseClause`.

A `UseClause` contains a symbol to use, and an optional limitation
clause containing some number of limitations. If a limitation clause
exists, it may be an `EXCEPT` list or an `ONLY` list.

The expressions in an `EXCEPT` list should only be `Identifier`. The
expressions in an `ONLY` list may be `Identifier` or `As` (they may
be renamed). This will have to be checked by the resolver.

Handle renaming in use statements by introducing the `As` node. It
contains an `Expression` called `symbol()`, which may be a `Dot` or an
`Identifier`. It also contains an `Identifier` called `rename()`.

Also adjust the `end()` bounds of several statement body iterators by
removing the `-1` at the end.

The `Import` node will be added in a separate effort.

Reviewed by @mppf, @lydia-duncan, @stonea. Thanks!

TESTING:

  - [x] All `compiler/next` tests pass
  - [x] Could build `chpl`

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Hi @lydia-duncan would you be the primary reviewer for this?
Hi @mppf would you mind checking the technicals? I moved the `contentsMatchInner` into the header this time :).

File: compiler/next/include/chpl/uast/ASTClassesList.h [hunk 1]
```diff
@@ -67,7 +67,9 @@ AST_BEGIN_SUBCLASSES(Expression)       // old AST: Expr
   //AST_NODE(Import)                     // old AST: ImportStmt
   AST_NODE(New)
   //AST_NODE(Require)                    //
-  //AST_NODE(Use)                        // old AST: UseStmt
+  AST_NODE(As)                         //
+  AST_NODE(Use)                        // old AST: UseStmt
+  AST_NODE(UseClause)                  //
```
Comment (lydia-duncan) [inline compiler/next/include/chpl/uast/ASTClassesList.h:72]: The chunks seem to usually be arranged alphabetically, but this breaks that.  Does this imply that Use and related nodes should end up in their own chunk?

File: compiler/next/test/frontend/testParseUse.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+  // Check the third limitation.
+  {
+    assert(useClause->limitation(2)->isAs());
+    const As* as = useClause->limitation(2)->toAs();
+    assert(as->name()->isIdentifier());
+    assert(as->name()->toIdentifier()->name() == "Baz");
+    assert(as->rename()->isIdentifier());
+    assert(as->rename()->toIdentifier()->name() == "Y");
+  }
+}
```
Comment (lydia-duncan) [inline compiler/next/test/frontend/testParseUse.cpp:229]: I don't see tests of `except *;` or `only;`.  I'm a little worried the check for limitations in the `only;` case is wrong.

**Review by lydia-duncan — COMMENTED** [review]
Nice!  Just a couple of concerns

File: compiler/next/include/chpl/uast/ASTClassesList.h (hunk 1 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/ASTClassesList.h:72]: Nope, that's just an error on my part.

File: compiler/next/test/frontend/testParseUse.cpp (hunk 2 above)
Comment (dlongnecke-cray) [reply compiler/next/test/frontend/testParseUse.cpp:229]: I will add some tests for those.

Can you explain what you mean (your fears) by your second sentence?

Comment (dlongnecke-cray) [reply compiler/next/test/frontend/testParseUse.cpp:229]: Oh I see. I can add a test for `only;`. I expect the limitations list will be empty and the `limitationClauseKind()` will be `UseClause::ONLY`.

File: compiler/next/include/chpl/uast/ASTClassesList.h (hunk 1 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/ASTClassesList.h:72]: Actually @mppf do we want to make an effort to arrange everything alphabetically? I'm actually kind of worried that the deep levels of nesting for decls will make things hard to follow (perhaps unfounded).

File: compiler/next/test/frontend/testParseUse.cpp (hunk 2 above)
Comment (dlongnecke-cray) [reply compiler/next/test/frontend/testParseUse.cpp:229]: Tests added, and fixed that bug. Thanks!

File: compiler/next/include/chpl/uast/UseClause.h [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  \rst
+  .. code-block:: chapel
+    // This contains two use clauses, 'Foo as X' and 'Baz as Y'.
+    use Foo as X, Baz as Y;
+
+  \endrst
+ */
+class UseClause final : public Expression {
+ public:
+  enum LimitationClauseKind {
```
Comment (stonea) [inline compiler/next/include/chpl/uast/UseClause.h:43]: `enum class` or `enum struct` is preferred. What do these values represent?

Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/UseClause.h:43]: I think the current strategy is to use old enums for things that are already namespaced by a class. The reason why I got away with using `enum class` for `BlockStyle` is because it lives outside of any one node and the values needed to be namespaced by _something_.

Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/UseClause.h:43]: I'll add a comment, thanks!

File: compiler/next/include/chpl/uast/ASTClassesList.h (hunk 1 above)
Comment (mppf) [reply compiler/next/include/chpl/uast/ASTClassesList.h:72]: The idea was to have the ones that aren't parent classes/groups in sorted order, followed by the parent classes/groups. The file is mostly already organized in this manner but there are probably a few other drifts away from it.

File: compiler/next/include/chpl/uast/UseClause.h (hunk 3 above)
Comment (mppf) [reply compiler/next/include/chpl/uast/UseClause.h:43]: Yeah. A reasonable alternative strategy would be to put all of these enums in a header file for the enums (rather than defining them within the class). Anyway it seems to me that `UseClause::LimitationClauseKind::EXCEPT` is a bit much for a long name so I wouldn't want to see both of those required.

File: compiler/next/include/chpl/uast/As.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ */
+class As final : public Expression {
+ private:
+  As(ASTList children)
+    : Expression(asttags::As, std::move(children)) {
+    const Expression* name = this->name();
+    assert(name->isIdentifier() || name->isDot()); 
+  }
+
+  // No need to mark 'nameChildNum_' or 'renameChildNum_', they are const.
```
Comment (mppf) [inline compiler/next/include/chpl/uast/As.h:50]: mark -> match

File: compiler/next/include/chpl/uast/As.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  const Expression* name() const {
+    auto ret = child(nameChildNum_);
+    assert(ret->isExpression());
+    return (const Expression*)ret;
+  }
+
+  /**
+    Return the rename specified by this as expression.
+  */
+  const Identifier* rename() const {
```
Comment (mppf) [inline compiler/next/include/chpl/uast/As.h:85]: I think this one could be represented as a UniqueString field and return a UniqueString. I'm not objecting to having an Identifier for it - I just don't really understand why we need to do that one way or the other.

(`name` is different because it might be a `Dot`).

File: compiler/next/include/chpl/uast/As.h [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    Create and return an as expression.
+  */
+  static owned<As> build(Builder* builder, Location loc,
+                         owned<Expression> name,
+                         owned<Identifier> rename);
+
+  /**
+    Return the original name specified by this as expression.
+  */
+  const Expression* name() const {
```
Comment (mppf) [inline compiler/next/include/chpl/uast/As.h:76]: I'm not really confident in the name of this method `name()` because we already use that in `NamedDecl` and `Identifier` as something that returns a `UniqueString`. It seems like it's breaking that pattern if this one returns an `Expression*`. But, if we can't think of a better `name()` this is OK :)

File: compiler/next/include/chpl/uast/Conditional.h [hunk 7]
```diff
@@ -135,7 +135,7 @@ class Conditional final : public Expression {
   */
   ASTListIteratorPair<Expression> thenStmts() const {
     auto begin = children_.begin() + thenBodyChildNum_;
-    auto end = begin + numThenBodyStmts_ - 1;
```
Comment (mppf) [inline compiler/next/include/chpl/uast/Conditional.h:138]: Looks like there are a few bug fixes here and in Loop - it'd be good to mention that the PR includes some bug fixes in the PR message.

File: compiler/next/include/chpl/uast/Use.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+      // Example 1:
+      use Foo, Bar as A;
+
+  \endrst
+
+  This creates a use statement that has two use clauses, 'Foo' and
+  'Bar as A'.
+*/
+class Use final : public Expression {
```
Comment (mppf) [inline compiler/next/include/chpl/uast/Use.h:46]: There is probably some reason you didn't do this but I thought we were considering having `Use` inherit from `Decl` ? Is your point that it's not really declaring anything (rather just making something available or possibly renaming something?). I think we could view them as declaring something but it's more about the exact meaning of "declaration" that we use than anything deeper.

Anyway I am not worried about it because the follow-on resolution steps will need to treat Use/Import directly since they affect what symbols are available so much.

File: compiler/next/include/chpl/uast/Use.h [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+ private:
+  Use(ASTList children, Decl::Visibility visibility)
+    : Expression(asttags::Use, std::move(children)),
+      visibility_(visibility) {
+    assert(numChildren() >= 1);
+    assert(isExpressionASTList(children_));
```
Comment (mppf) [inline compiler/next/include/chpl/uast/Use.h:52]: Child expressions can only be `UseClause`, right? Can you add a function to call in an assert to check that? I have added such a pattern several times, see e.g. https://github.com/chapel-lang/chapel/blob/master/compiler/next/include/chpl/uast/MultiDecl.h#L59

File: compiler/next/include/chpl/uast/UseClause.h [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+  void markUniqueStringsInner(Context* context) const override {
+    expressionMarkUniqueStringsInner(context);
+  }
+
+  // This always exists and its position never changes.
+  static const int8_t nameChildNum_ = 0;
+
+  LimitationClauseKind limitationClauseKind_;
+  int8_t limitationClauseChildNum_;
+  int numLimitations_;
```
Comment (mppf) [inline compiler/next/include/chpl/uast/UseClause.h:83]: It's not a big deal but I tend to prefer to see the fields immediately after the types declared (so it would be right before the constructor)

File: compiler/next/include/chpl/uast/UseClause.h [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  static owned<UseClause> build(Builder* builder, Location loc,
+                                owned<Expression> name,
+                                LimitationClauseKind limitationClauseKind,
+                                ASTList limitationClause);
+
+  /**
+    Get the name expression of this use clause. It may be a dot expression,
+    an as expression, or an identifier.
+  */
+  const Expression* name() const {
```
Comment (mppf) [inline compiler/next/include/chpl/uast/UseClause.h:106]: See my other comment about `name()`

File: compiler/next/include/chpl/uast/UseClause.h [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+  void markUniqueStringsInner(Context* context) const override {
+    expressionMarkUniqueStringsInner(context);
+  }
+
+  // This always exists and its position never changes.
+  static const int8_t nameChildNum_ = 0;
+
+  LimitationClauseKind limitationClauseKind_;
+  int8_t limitationClauseChildNum_;
```
Comment (mppf) [inline compiler/next/include/chpl/uast/UseClause.h:82]: Isn't this always 1 ? Seems like you could just check `numLimitations > 0` and treat this one as 1 in that event.

File: compiler/next/include/chpl/uast/UseClause.h [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+  */
+  LimitationClauseKind limitationClauseKind() const {
+    return limitationClauseKind_;
+  }
+
+  /**
+    Return a way to iterate over the limitations of this use clause.
+  */
+  ASTListIteratorPair<Expression> limitations() const {
+    auto begin = numLimitations()
```
Comment (mppf) [inline compiler/next/include/chpl/uast/UseClause.h:123]: can we use `numLimitations() > 0` ?

File: compiler/next/include/chpl/uast/UseClause.h [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  /**
+    Return the i'th limitation of this use clause. If the limitation clause
+    kind is 'EXCEPT', then the limitations will all be identifiers. If the
+    limitation clause kind is 'ONLY', then the limitations may be
+    identifers or as expressions containing identifiers.
+  */
+  const Expression* limitation(int i) const {
+    if (limitationClauseChildNum_ < 0) return nullptr;
```
Comment (mppf) [inline compiler/next/include/chpl/uast/UseClause.h:144]: Seems like maybe this should be an assert too.

File: compiler/next/lib/frontend/Parser/ParserContextImpl.h [hunk 15]
```diff
@@ -1083,3 +1083,47 @@ Expression* ParserContext::buildNumericLiteral(YYLTYPE location,
...
+Expression* ParserContext::buildAsExpr(YYLTYPE locName, YYLTYPE locRename,
+                                       owned<Expression> name,
+                                       owned<Expression> rename) {
+  if (!rename->isIdentifier()) {
+    const char* msg = "Rename in as expression must be identifier";
+    return raiseError(locRename, msg);
+  }
+
+  auto renameAsIdent = toOwned(rename.release()->toIdentifier());
+  assert(renameAsIdent.get());
```
Comment (mppf) [inline compiler/next/lib/frontend/Parser/ParserContextImpl.h:1096]: Not complaining but this assert seems redundant with the above check.

**Review by mppf — APPROVED** [review]

File: compiler/next/include/chpl/uast/Conditional.h (hunk 7 above)
Comment (lydia-duncan) [reply compiler/next/include/chpl/uast/Conditional.h:138]: I think they already are (at least, that's why I remember not calling it out myself)

Comment (mppf) [reply compiler/next/include/chpl/uast/Conditional.h:138]: Indeed they are! Thanks for pointing out that I had missed them.

> Also adjust the end() bounds of several statement body iterators by
removing the -1 at the end.

**Review by lydia-duncan — APPROVED** [review]
All my issues have been resolved, thanks!

**Review by stonea — APPROVED** [review]

File: compiler/next/include/chpl/uast/As.h (hunk 5 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/As.h:85]: It seems like we'd want to have identifiers so that we can have precise line numbers.

File: compiler/next/include/chpl/uast/As.h (hunk 6 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/As.h:76]: How about something like `symbol()`?

File: compiler/next/include/chpl/uast/Use.h (hunk 8 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/Use.h:46]: I ended up not doing that because it left more questions than answers.

File: compiler/next/include/chpl/uast/As.h (hunk 6 above)
Comment (mppf) [reply compiler/next/include/chpl/uast/As.h:76]: That sounds nice to me! I also like that it makes the potential `name` vs `rename` confusion less.

File: compiler/next/include/chpl/uast/As.h (hunk 5 above)
Comment (mppf) [reply compiler/next/include/chpl/uast/As.h:85]: Not sure how much that'll matter in practice but it's OK this way to start, anyway.

File: compiler/next/include/chpl/uast/UseClause.h (hunk 10 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/UseClause.h:83]: I will make a cleanup TODO list

File: compiler/next/include/chpl/uast/ASTClassesList.h (hunk 1 above)
Comment (dlongnecke-cray) [reply compiler/next/include/chpl/uast/ASTClassesList.h:72]: I sorted everything but the abstract class "chunks".
